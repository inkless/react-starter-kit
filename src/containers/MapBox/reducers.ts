import { getType } from 'typesafe-actions';

import { MapboxState } from './types';
import {
  MapboxAction,
  fetchOriginGeocodingAction,
  setViewport,
  setDeliverSimulationStarted,
  setDeliverGeoFeatures,
  moveCamera,
} from './actions';
import { Feature, Point } from '@turf/helpers';

const defaultState: MapboxState = {
  origin: {
    address: '',
    coordinates: [0, 0],
  },
  viewport: {
    longitude: 0,
    latitude: 0,
    zoom: 11.5,
  },
  deliverSimulationStarted: false,
  deliverSimulationGeoFeatures: [],
};

export function mapboxReducer(state = defaultState, action: MapboxAction): MapboxState {
  switch (action.type) {
    case getType(fetchOriginGeocodingAction.success): {
      const [longitude, latitude] = action.payload.coordinates;
      return {
        ...state,
        origin: action.payload,
        viewport: {
          ...state.viewport,
          longitude,
          latitude,
        },
      };
    }
    case getType(setViewport): {
      // Note: this is needed, ReactMapGL will send width and height with the
      // viewport, in order to allow map resize properly when window resize,
      // we need to remove this width, height from payload
      const { width, height, ...viewport } = action.payload;
      return {
        ...state,
        viewport,
      };
    }
    case getType(setDeliverSimulationStarted):
      return {
        ...state,
        deliverSimulationStarted: action.payload,
      };
    case getType(setDeliverGeoFeatures):
      return {
        ...state,
        deliverSimulationGeoFeatures: action.payload,
      };
    case getType(moveCamera): {
      const courierPoint = state.deliverSimulationGeoFeatures.find(
        feature => feature.properties.id === action.payload
      ) as Feature<Point, any> | undefined;

      if (!courierPoint || !courierPoint.geometry) {
        return state;
      }
      const [longitude, latitude] = courierPoint.geometry.coordinates;
      return {
        ...state,
        viewport: {
          ...state.viewport,
          longitude,
          latitude,
        },
      };
    }
    default:
      return state;
  }
}
