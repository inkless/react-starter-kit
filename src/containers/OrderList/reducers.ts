import { OrderListState } from './types';
import {
  OrderListAction,
  updatePendingOrders,
  startDeliveringOrder,
  orderDelivered,
  setSimulatingOrdersStarted,
} from './actions';
import { getType } from 'typesafe-actions';

const defaultState: OrderListState = {
  pending: [],
  delivering: [],
  delivered: [],
  simulatingOrdersStarted: false,
};

export function orderListReducer(state = defaultState, action: OrderListAction): OrderListState {
  switch (action.type) {
    case getType(updatePendingOrders):
      return {
        ...state,
        pending: action.payload,
      };
    case getType(startDeliveringOrder):
      return {
        ...state,
        delivering: [action.payload, ...state.delivering],
      };
    case getType(orderDelivered):
      return {
        ...state,
        // remove from delivering using the delivered order's id
        delivering: state.delivering.filter(({ id }) => id !== action.payload.id),
        delivered: [action.payload, ...state.delivered],
      };
    case getType(setSimulatingOrdersStarted):
      return {
        ...state,
        simulatingOrdersStarted: action.payload,
      };
    default:
      return state;
  }
}
