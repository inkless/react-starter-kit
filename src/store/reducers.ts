import { connectRouter } from 'connected-react-router';
import { combineReducers } from 'redux';
import { History } from 'history';

import { mapboxReducer } from '@src/containers/MapBox/reducers';
import { orderListReducer } from '@src/containers/OrderList/reducers';

export const createRootReducer = (history: History) =>
  combineReducers({
    router: connectRouter(history),
    mapbox: mapboxReducer,
    orderList: orderListReducer,
  });
