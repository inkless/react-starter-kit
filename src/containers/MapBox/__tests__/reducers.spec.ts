import { fixtureData1 } from './__fixture__/geojson';
import {
  fetchOriginGeocodingAction,
  setViewport,
  setDeliverSimulationStarted,
  setDeliverGeoFeatures,
} from '../actions';
import { mapboxReducer } from '../reducers';

describe('MapBox Reducers', () => {
  it('should reduce correct state for fetchOriginGeocodingAction', () => {
    const action = fetchOriginGeocodingAction.success({
      address: 'Home',
      coordinates: [-137, 37],
    });
    expect(mapboxReducer(undefined, action)).toMatchSnapshot();
  });

  it('should reduce correct state for setViewport', () => {
    const action = setViewport({
      longitude: -137,
      latitude: 37,
      zoom: 11,
    });
    expect(mapboxReducer(undefined, action)).toMatchSnapshot();
  });

  it('should reduce correct state for setDeliverSimulationStarted', () => {
    const action = setDeliverSimulationStarted(true);
    expect(mapboxReducer(undefined, action)).toMatchSnapshot();
  });

  it('should reduce correct state for setDeliverGeoFeatures', () => {
    const action = setDeliverGeoFeatures([fixtureData1.route]);
    expect(mapboxReducer(undefined, action)).toMatchSnapshot();
  });
});
