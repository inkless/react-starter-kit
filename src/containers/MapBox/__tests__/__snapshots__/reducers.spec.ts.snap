// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapBox Reducers should reduce correct state for fetchOriginGeocodingAction 1`] = `
Object {
  "deliverSimulationGeoFeatures": Array [],
  "deliverSimulationStarted": false,
  "origin": Object {
    "address": "Home",
    "coordinates": Array [
      -137,
      37,
    ],
  },
  "viewport": Object {
    "latitude": 37,
    "longitude": -137,
    "zoom": 11.5,
  },
}
`;

exports[`MapBox Reducers should reduce correct state for setDeliverGeoFeatures 1`] = `
Object {
  "deliverSimulationGeoFeatures": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            -118.46742,
            34.00534,
          ],
          Array [
            -118.47105,
            34.00321,
          ],
          Array [
            -118.47115,
            34.00309,
          ],
          Array [
            -118.47325,
            34.00463,
          ],
          Array [
            -118.47629,
            34.00289,
          ],
          Array [
            -118.48058,
            33.99993,
          ],
          Array [
            -118.48113,
            33.99954,
          ],
          Array [
            -118.48088,
            33.9993,
          ],
          Array [
            -118.48088,
            33.99928,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {},
      "type": "Feature",
    },
  ],
  "deliverSimulationStarted": false,
  "origin": Object {
    "address": "",
    "coordinates": Array [
      0,
      0,
    ],
  },
  "viewport": Object {
    "latitude": 0,
    "longitude": 0,
    "zoom": 11.5,
  },
}
`;

exports[`MapBox Reducers should reduce correct state for setDeliverSimulationStarted 1`] = `
Object {
  "deliverSimulationGeoFeatures": Array [],
  "deliverSimulationStarted": true,
  "origin": Object {
    "address": "",
    "coordinates": Array [
      0,
      0,
    ],
  },
  "viewport": Object {
    "latitude": 0,
    "longitude": 0,
    "zoom": 11.5,
  },
}
`;

exports[`MapBox Reducers should reduce correct state for setViewport 1`] = `
Object {
  "deliverSimulationGeoFeatures": Array [],
  "deliverSimulationStarted": false,
  "origin": Object {
    "address": "",
    "coordinates": Array [
      0,
      0,
    ],
  },
  "viewport": Object {
    "latitude": 37,
    "longitude": -137,
    "zoom": 11,
  },
}
`;
